### PR TITLE
Card origin_devexp_286 - Added ssh key-based access to private git repository documentation

### PR DESCRIPTION
**Pull request by soltysh**
@bparees @CowboysFan @mfojtik PTAL

### CODE REVIEW

**Comment by adellape** [comment]
Related docs card: https://trello.com/c/nsFhtpIV

File: architecture/core_objects/builds.adoc [hunk 1]
```diff
@@ -164,3 +165,68 @@ $ osc create -f secret.json
 Once you have the `*Secret*` created, you can add a `PushSecretName` field into
 `Output` section inside the `BuildConfig` and set it to the name of the
 `*Secret*` that you created, in this case `*dockerhub*`.
+
+[#using-private-repositories-for-builds]
+== Using private repositories for builds
+
+In case you want to build your application from a private repository you must
```
Comment (bparees) [inline architecture/core_objects/builds.adoc]: s/In case you/If you/

File: architecture/core_objects/builds.adoc [hunk 2]
```diff
@@ -164,3 +165,68 @@ $ osc create -f secret.json
...
+supply a valid credentials, currently only SSH key based authentication is supported.
```
Comment (bparees) [inline architecture/core_objects/builds.adoc]: s/supply a valid/supply valid/
s/credentials, currently/credentials.  Currently/

File: architecture/core_objects/builds.adoc [hunk 3]
```diff
@@ -164,3 +165,68 @@ $ osc create -f secret.json
...
+The keys used to access that repository lives in `$HOME/.ssh/` directory.
```
Comment (bparees) [inline architecture/core_objects/builds.adoc]: s/keys/key/ or s/lives/live/

File: architecture/core_objects/builds.adoc [hunk 4]
```diff
@@ -164,3 +165,68 @@ $ osc create -f secret.json
...
+If you don't have any, you can generate them with the following command:
+
+====
+
+----
+$ ssh-keygen -t rsa -C "your_email@example.com"
+----
+====
+
+Once that's done you're gonna get two files: one containing your public key (as
```
Comment (bparees) [inline architecture/core_objects/builds.adoc]: s/gonna/going to/

File: architecture/core_objects/builds.adoc [hunk 5]
```diff
@@ -164,3 +165,68 @@ $ osc create -f secret.json
...
+explained above) and corresponding private key (one of `id_dsa`, `id_ecdsa`,
```
Comment (bparees) [inline architecture/core_objects/builds.adoc]: s/and corresponding/and one containing a corresponding/

File: architecture/core_objects/builds.adoc [hunk 6]
```diff
@@ -164,3 +165,68 @@ $ osc create -f secret.json
...
+`id_ed25519` or `id_rsa`). With both of these in place you should consult your
+SCM manual on how to upload the public key, the private one will be then used to
```
Comment (bparees) [inline architecture/core_objects/builds.adoc]: s/SCM/source control management(SCM) system's manual/

Comment (bparees) [inline architecture/core_objects/builds.adoc]: s/public key, the private/public key.  The private/

File: architecture/core_objects/builds.adoc [hunk 7]
```diff
@@ -164,3 +165,68 @@ $ osc create -f secret.json
...
+access your private repository.
+
+Kubernetes provides the
+https://github.com/GoogleCloudPlatform/kubernetes/blob/master/docs/design/secrets.md[Secret]
+resource, which you can use to store your keys. You must create the `*Secret*` first
+before builds can use your ssh key for accessing your private repository.
+The `*data*` field for the `*Secret*` object must contain your private key
+with the value set to the base64-encoded content of that file. For example:
```
Comment (bparees) [inline architecture/core_objects/builds.adoc]: probably worth giving a full end to end example here... ie show examples/commands/json for how to convert a keyfile to base64, put the value into a Secret json definition, create the secret in openshift, reference the secret in a build.

**Comment by bparees** [comment]
couple minor fixes and one medium size suggestion, otherwise looks good!

**Comment by soltysh** [comment]
@bparees addressed all your comments.

File: architecture/core_objects/builds.adoc [hunk 8]
```diff
@@ -164,3 +165,79 @@ $ osc create -f secret.json
...
+
+====
+
+----
+$ base64 -w 0 $HOME/.ssh/id_rsa
+6yJodHRwc1ovL2zuZGV4LmRvY21lci5aby92MS8iOnsiYXV0aCI6ImJXWnZhblJwYXpwdVoybGxkR2d4TUE9PSIsImVtYWlsIj8ibWlAbWlmby5zayJ9fQ==
+----
+====
+
+Copy the value returned from `base64` command and paste it into `sshsecert` field
```
Comment (bparees) [inline architecture/core_objects/builds.adoc]: s/sshsecert/sshsecret/

File: architecture/core_objects/builds.adoc [hunk 9]
```diff
@@ -164,3 +165,79 @@ $ osc create -f secret.json
...
+  },
+  "data": {
+    "sshsecret": "6yJodHRwc1ovL2zuZGV4LmRvY21lci5aby92MS8iOnsiYXV0aCI6ImJXWnZhblJwYXpwdVoybGxkR2d4TUE9PSIsImVtYWlsIj8ibWlAbWlmby5zayJ9fQ=="
+  }
+}
+
+----
+====
+
+To create the `*Secret*` from a *_secret.json_* file, for example, you can use
```
Comment (bparees) [inline architecture/core_objects/builds.adoc]: "You can then create the `*Secret*` from the *_secret.json_* file using the following command:"

File: architecture/core_objects/builds.adoc [hunk 10]
```diff
@@ -164,3 +165,79 @@ $ osc create -f secret.json
...
+====
+
+----
+$ osc create -f secret.json
+----
+====
+
+Once you have the `*Secret*` created, you can add a `SourceSecretName` field into
+`Source` section inside the `BuildConfig` and set it to the name of the
+`*Secret*` that you created, in this case `*scmsecret*`.
```
Comment (bparees) [inline architecture/core_objects/builds.adoc]: one more example json here, for the BuildConfig.

**Comment by bparees** [comment]
couple more comments.

File: architecture/core_objects/builds.adoc (hunk 8 above)
Comment (soltysh) [inline architecture/core_objects/builds.adoc]: It'll be changed to `ssh-privatekey` based on discussion [here](https://github.com/openshift/origin/pull/1926/files#r29672163).

Comment (adellape) [inline architecture/core_objects/builds.adoc]: @soltysh Just to confirm, when `sshsecret` is changed to `ssh-privatekey`, will `ssh-privatekey` be specifically what is required for this, like the `dockercfg` key in the previous section? I'm assuming so based on the "Supported auth methods are: ssh-privatekey" comments in the linked dev PR.

Comment (soltysh) [inline architecture/core_objects/builds.adoc]: Yes that's correct, based on the key name we'll choose the appropriate auth method. For now ssh-privatekey is the only supported, future versions will support user+password and token etc. I'll update that PR as soon as my code PR is done, which I'm pretty close to.

**Comment by soltysh** [comment]
@bparees updated & addressed your comments. PTAL.

**Comment by bparees** [comment]
looks good to me, excellent work @soltysh!

**Comment by soltysh** [comment]
@bparees I have good reviewers :smile:

**Comment by adellape** [comment]
@soltysh I'm assuming I should wait til the code PR merges before merging this. I'll slap a [do not merge] on this until told otherwise. Thanks!

**Comment by bparees** [comment]
yeah, you'll want to wait on this PR: https://github.com/openshift/origin/pull/1926

**Comment by soltysh** [comment]
@adellape the PR @bparees mentioned is merged, you can safely merge this one as well.

**Pull request merged**